src/RequestFlags.h

```diff
@@ -28,8 +30,10 @@ class RequestFlags
     bool auth = false;
     /** do not use keytabs for peer Kerberos authentication */
     bool auth_no_keytab = false;
-    /** he response to the request may be stored in the cache */
-    bool cachable = false;
+
+    /// whether the response may be stored in the cache
+    SupportOrVeto cachable;
+
```

Comment:
> Amos at https://github.com/squid-cache/squid/pull/1185#discussion_r1032642857: Correctly implemented this request flag should mirror the HTTP semantics with default-true for easier validation of the resulting Squid behaviour. 

I strongly disagree. Due to the security concerns related to placing content in a _shared_ cache, we must continue to default this flag to false. While going against HTTP's "reuse is the default" design principle might feel counter-intuitive, it is the best implementation approach (among the flag-based ones) because not all flagged traffic is HTTP, and because there are no "MUST cache" protocol requirements (while there are plenty of "MUST NOT cache" rules that we must check if we decide to cache).

Caching is and will remain an _optional_ security-sensitive optimization. As long as we limit ourselves to a simple boolean decision flag, the only safe implementation default is "false". Thankfully, it is also the current default.

----

It would be even better to guarantee that the code has reached all the required caching decision points for the involved protocol(s) before allowing caching. In other words, the ideal implementation would upgrade from the boolean support-or-veto flag (that may miss a required decision making point and not know it) to an **all**-the-required-decision-makers-support-or-veto state. Reaching that ideal is too complex for this PR, but this PR changes will make such future adjustments (if any) easier by correctly enforcing vetos (and preserving the right default).

Reply:
You are mixing concepts here. Caching as a whole is optional, that is not related to whether an individual response default is to *be* cacheable.

From a security standpoint the risks are greater whenever the standard behaviour is *not* followed. HTTP/1.1 has clear mechanisms which Squid makes use of now to prevent casual risks from cached security sensitive content (eg private responses, corrupt responses) affecting the network. This ```cacheable``` flag is not part of those protections, in part because it does not follow the standards and thus cannot be relied on.
  It would be nice to make it properly compliant with HTTP so we *could* rely on seeing a ```false``` value for security checks to forbid caching. Right now we cannot trust false to be true or just the HTTP-violation default.

Reply:
> Caching as a whole is optional, that is not related to whether an individual response default is to *be* cacheable.

Agreed. Nothing in my earlier response or this PR contradicts the above statement AFAICT.

> From a security standpoint the risks are greater whenever the standard behaviour is not followed.

In the scope of this PR, Squid implementation continues to follow the standard by refusing to cache objects that have not been through any standard cachability verification rules. The standard requires such verification (because there are MUST NOT cache requirements that a compliant implementation ought to check/obey). The official and PR code enforces that basic principle.

> HTTP/1.1 has clear mechanisms which Squid makes use of now to prevent casual risks from cached security sensitive content (eg private responses, corrupt responses) affecting the network.

Yes, and when Squid code implementing those mechanisms is reached, and request cachability is verified (i.e. the risks are mitigated), the cachable flag should become true. Until those mechanisms are reached, the flag should remain false. The official and PR code use that approach. It is not only very reasonable, it is probably the only safe approach based on a single boolean flag (as I detailed earlier).

> This cacheable flag is not part of those protections,

I disagree: We set this flag to true when HttpRequest::maybeCacheable() returns true. We veto caching when HttpRequest::maybeCacheable() returns false. That HttpRequest method is a part of those protections and, hence, this flag is too.

> It would be nice to make it properly compliant with HTTP so we _could_ rely on seeing a `false` value for security checks to forbid caching. 

We can (and do) already "rely on seeing a `false` value for security checks to forbid caching". No changes to the flag default are needed to achieve that goal! A false value means that we have not checked certain MUSTs yet or we have checked them and they prohibited caching. In both cases, we should forbid caching.

This flag does have a problem, but it is not its default value. It is its lack of awareness of whether we have checked all MUSTs that we are supposed to check. That problem cannot be solved within the confines of a boolean flag (as I detailed earlier).

> Right now we cannot trust false to be reality or just the HTTP-violation default.

Agreed (although it does not violate any HTTP MUSTs). And if we change the default from false to true, we would not be able to trust `true` to be reality either! To be able to fully trust this field, we need to convert it from a boolean flag to a list of the required protocol-specific checkers (that are activated later in the code). Only then we can really trust this field to fully reflect reality.